File: packages/python/pii_vault/providers/anthropic.py
from typing import Iterator, Optional

from ..detector import Detector
from ..tokenizer import Tokenizer
from ..vault import LocalVault


class SafeAnthropic:
    """
    Drop-in replacement for anthropic.Anthropic.

    Local vault (default)::

        client = SafeAnthropic(api_key="sk-ant-...", vault_key="my-secret")

    Hosted vault (Mawlaia cloud)::

        client = SafeAnthropic(
            api_key="sk-ant-...",
            vault_key="my-hmac-secret",
            mawlaia_api_key="mwl_live_...",
        )
    """

    def __init__(
        self,
        api_key:          str,
        vault_key:        str = "",
        vault_path:       str = ":memory:",
        mawlaia_api_key:  Optional[str] = None,
        mawlaia_vault_url: str = "https://api.mawlaia.com",
        vault=None,
        entities:         Optional[list[str]] = None,
        llm_fallback:     bool = False,
        token_mode:       str = "typed",
        **anthropic_kwargs,
    ):
        from anthropic import Anthropic

        self._raw = Anthropic(api_key=api_key, **anthropic_kwargs)

        if vault is not None:
            _vault = vault
        elif mawlaia_api_key:
            from ..hosted_vault import HostedVault
            _vault = HostedVault(api_key=mawlaia_api_key, vault_url=mawlaia_vault_url)
        else:
            _vault = LocalVault(path=vault_path)

        detector  = Detector(entities=entities, llm_fallback=llm_fallback, openai_api_key=None)
        tokenizer = Tokenizer(vault=_vault, key=vault_key or "default", mode=token_mode)

        self.messages = _MessagesNamespace(self._raw, detector, tokenizer)

    def __getattr__(self, name: str):
        return getattr(self._raw, name)


class _MessagesNamespace:
    def __init__(self, client, detector: Detector, tokenizer: Tokenizer):
        self._client    = client
        self._detector  = detector
        self._tokenizer = tokenizer

    def create(self, *, messages: list[dict], subject_id: Optional[str] = None, **kwargs):
        tokenized = self._tokenizer.tokenize_messages(
            messages, self._detector, subject_id=subject_id
        )

        if kwargs.get("stream"):
            return _AnthropicStream(
                self._client.messages.create(messages=tokenized, **kwargs),
                self._tokenizer,
            )

        response = self._client.messages.create(messages=tokenized, **kwargs)

        # Anthropic response: response.content is a list of ContentBlock
        for block in response.content:
            if hasattr(block, "text") and block.text:
                block.text = self._tokenizer.dehydrate(block.text)

        return response


class _AnthropicStream:
    """
    Wraps an Anthropic streaming response and dehydrates text deltas.
    """

    def __init__(self, stream, tokenizer: Tokenizer):
        self._stream    = stream
        self._tokenizer = tokenizer
        self._buffer    = ""

    def __iter__(self) -> Iterator:
        for event in self._stream:
            # Anthropic streams text_delta events
            if hasattr(event, "delta") and hasattr(event.delta, "text"):
                self._buffer += event.delta.text
                safe, self._buffer = self._tokenizer.split_stream_safe(self._buffer)
                if safe:
                    event.delta.text = self._tokenizer.dehydrate(safe)
                    yield event
            else:
                yield event

        if self._buffer:
            import copy
            final = copy.deepcopy(event)  # type: ignore[possibly-unbound]
            final.delta.text = self._tokenizer.dehydrate(self._buffer)
            self._buffer = ""
            yield final

    def __enter__(self):
        return self

    def __exit__(self, *args):
        if hasattr(self._stream, "__exit__"):
            self._stream.__exit__(*args)

File: packages/typescript/src/providers/anthropic.ts
import { Detector } from "../detector";
import { Message } from "../models";
import { Tokenizer } from "../tokenizer";
import { Vault } from "../vault";

export interface SafeAnthropicOptions {
  apiKey:     string;
  vaultKey:   string;
  vaultPath?: string;
  entities?:  string[];
  tokenMode?: "typed" | "opaque";
  [key:       string]: unknown;
}

export class SafeAnthropic {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private raw:       any;
  private detector:  Detector;
  private tokenizer: Tokenizer;
  public  messages:  MessagesNamespace;

  constructor(options: SafeAnthropicOptions) {
    const { apiKey, vaultKey, vaultPath = ":memory:", entities, tokenMode = "typed", ...rest } = options;

    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { default: Anthropic } = require("@anthropic-ai/sdk");
    this.raw       = new Anthropic({ apiKey, ...rest });
    const vault    = new Vault(vaultPath);
    this.detector  = new Detector({ entities });
    this.tokenizer = new Tokenizer(vault, vaultKey, tokenMode);
    this.messages  = new MessagesNamespace(this.raw, this.detector, this.tokenizer);
  }
}

class MessagesNamespace {
  constructor(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private client:    any,
    private detector:  Detector,
    private tokenizer: Tokenizer,
  ) {}

  async create({
    messages,
    subjectId,
    ...kwargs
  }: { messages: Message[]; subjectId?: string; [key: string]: unknown }) {
    const tokenized = this.tokenizer.tokenizeMessages(messages, this.detector, subjectId);

    if (kwargs["stream"]) {
      const stream = await this.client.messages.create({ messages: tokenized, ...kwargs });
      return new AnthropicStream(stream, this.tokenizer);
    }

    const response = await this.client.messages.create({ messages: tokenized, ...kwargs });

    // Anthropic: response.content is ContentBlock[]
    for (const block of response.content ?? []) {
      if (block.type === "text" && block.text) {
        block.text = this.tokenizer.dehydrate(block.text as string);
      }
    }

    return response;
  }
}

class AnthropicStream {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(private stream: any, private tokenizer: Tokenizer) {}

  async *[Symbol.asyncIterator]() {
    let buffer = "";

    for await (const event of this.stream) {
      const text: string | undefined = event.delta?.text;

      if (text) {
        buffer += text;
        const [safe, remainder] = this.tokenizer.splitStreamSafe(buffer);
        buffer = remainder;
        if (safe) {
          event.delta.text = this.tokenizer.dehydrate(safe);
          yield event;
        }
      } else {
        yield event;
      }
    }

    if (buffer) {
      yield { type: "content_block_delta", delta: { type: "text_delta", text: this.tokenizer.dehydrate(buffer) } };
    }
  }
}
